sphinx/developers/tool-developer-guide.rst

```diff
@@ -0,0 +1,441 @@
+supports all of this. Thankfully this is not necessary.
+
+We are including support for the notion of "conformance levels" or
+"profiles" to LinkML. This will allow applications to clearly state
+what parts of the specification are supported, and for people
+deploying the application to adapt accordingly.
+
+.note:: An example use case is a simple questionnaire UI/application that is
+        driven by a LinkML schema, where each question is a slot and
+        each answer either free text or in the case of multiple
+        choice, an enum. There is no need for inheritance in this
+        case, so the application only needs to pay attention to
+        elements in the schema such as `slots` and
+        `permissible_values`.
+
+Currently specific conformance levels in LinkML have not been defined,
+but current thinking is to have different categories:
+
+- A *structural* conformance level that pertains to logical parts of
+  the schema
+
+    * The minimal level would include `classes`, `attributes`, `range`
+      and basic builtin types, as well as cardinalities
+    * The next level would include class inheritance and slots, and
+      basic enums
+    * The level above this would include slot inheritance, and slot
+      overrides with slot_usage
+    * The highest level would include boolean expressions and rules
+
+- A *metadata* conformance level that pertains to informational
+  aspects of the schema that do not drive validation or inference
+
+    * minimal metadata would include metaslots for `description`,
+      `comments`
+    * above this, a minimal semantic level would include class and
+      slot uris, and mappings
+    * the full level would include all possible metaslots
+
+The LinkML framework provides some functionality for being able to do
+transformations "down" levels. For example, a rich schema that uses
+inheritance and slot_usage can be "rolled down" to a schema that
+includes only leaf elements, with inferred slots materialized as
+attributes. The resulting schema will be less rich, but may be
+sufficient for certain kinds of applications.
+
+See the `YAML Generator <../generators/yaml>_` for one such tool
+
+Other Languages
+---------------
+
+LinkML is programming language neutral. However, currently much of the
+stack for introspecting schemas is written in Python.
+
+For example, in Pythonif you want a list of slots for that class you can use `class_slots
+<manipulating-schemas.html#linkml_runtime.utils.schemaview.SchemaView.class_slots>_`. But
+what if you are building a generic javascript data entry widget that
+will work for any LinkML class?
+
+There are a few broad approaches here:
+
+- 1. Use the direct output of your language's YAML/JSON parser, and
+  implement logic ad-hoc
+- 2. Develop a full blown analog of the python SchemaView class for
+  your language
+- 3. Write python code to transform LinkML YAML into a native form
+  geared for your application
+
```

Comment:
I think there is a fourth option, which we're doing in DataHarmonizer: Use LinkML python SchemaView and other calls to create a compiled data structure which can be read as YAML or JSON, which is then read-in by your platform and language of choice (e.g. javascript in a browser).